nixos-option C++ rewrite with --all

chkno opened the pull request:
##### Motivation for this change
I wanted a --all flag for nixos-option.  As a shell script, this took tens of hours to run and felt brittle.  As a direct client of the C++ nix library, it takes six seconds & feels robust.  Looking at diffs of the --all output before and after making changes in configuration.nix is amazingly helpful in understanding what's going on inside and for doing refactoring.

This implementation's output format is slightly different, and where different it is mostly an improvement (eg: indented).  I didn't try to exactly match the shell script's output format, but I am happy to pursue that if there are important automated consumers that cannot be atomically updated.

The shell script says it has an XML output mode, but it seems to do very little (it only works on non-leaf paths -- `sound` but not `sound.enable`), so I made no attempt to preserve XML output mode.

###### Things done

(I'm not sure which of these things make sense, how to do some of them, or what other testing would be useful.  I'm new to Nix.  Thanks for your help!)

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @nbp @Ma27

chkno committed:
nixos/nixos-option: Rewrite in a more suitable language

Also add --all, which shows the value of all options.  Diffing the --all
output on either side of contemplated changes is a lovely way to better
understand what's going on inside nixos.

chkno committed:
nixos/nixos-option: Show options' types #27920

chkno commented:
(To make review easier: I'm not sure what the normal way to build packages inside nixos is, but I found that I could build this with this invocation:
```
nix-build -E '((import <nixpkgs> {}).nixos {}).nixos-option' -I nixpkgs="$HOME/devel/nixpkgs"
```
)

Mic92 commented:
This should be supported by the @NixOS/nix-core team as we would need to keep `nixos-option` compatible with the nix API.

chkno committed:
nixos/nixos-option: Fix references to old name

edolstra commented:
For the C++ code, could you use the same formatting as Nix (e.g. 4 space indent)? See .dir-locals in the Nix source tree.

:+1: on dropping the XML option. It would make more sense to have a `--json` flag now.

edolstra commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,588 @@
+      out << v;
+    }
+  } catch (Error &e) {
+    if (e.msg() == "The option `" + path + "' is used but not defined.") {
```
Hm, matching on an error message is a bad idea. Maybe introduce a new exception type?

chkno replied:
I refined this from Error to ThrownError, but I don't know how to refine it further.  This is an error thrown from nix code (devel/nixpkgs/lib/modules.nix:329) not a place inside the nix interpreter where it would be straightforward to add another error type.

chkno replied:
I refined this from Error to ThrownError, but I don't know how to refine it further. This is an error thrown from nix code (devel/nixpkgs/lib/modules.nix:329) not a place inside the nix interpreter where it would be straightforward to add another error type.

edolstra commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,588 @@
+                                 *ctx->autoArgs, v));
+      if (getSubOptions.type != tLambda) {
+        throw Error("Option's type.getSubOptions isn't a function at '" + attr +
+                    "' in path '" + path + "'");
```
Please use format strings instead of concatenation, e.g.
```
throw Error("option's 'type.getSubOptions' is not a function at '%s' in path '%s'", attr, path);
```

edolstra commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,588 @@
+
+bool shouldSort(Context *ctx, Value &v) {
+  // Some lists should clearly be printed in sorted order, like
+  // environment.systemPackages.  Some clearly should not, like
```
Hm, I don't think *any* lists should be sorted. E.g. the order of `environment.systemPackages` matters with respect to file collisions.

edolstra commented:
@chkno The canonical way to build NixOS-specific tools is like this:
```
nix-build '<nixpkgs/nixos>' -A config.system.build.nixos-install
```

chkno committed:
Reformat for 4-space indentation

Specifically, with
  clang-format --style='{ IndentWidth: 4, BreakBeforeBraces: Mozilla, ColumnLimit: 120, PointerAlignment: Middle }'
which was the clang-format invocation that produced the fewest diffs on
the nix source out of ~20 that I tried.

chkno committed:
Use format strings, not concatenation, in error messages

chkno committed:
Remove list sorting

chkno committed:
Parallel build is the default, so no need to specify

FRidh commented on nixos/modules/installer/tools/nixos-option/default.nix:
```diff
@@ -0,0 +1,7 @@
+  src = ./.;
+  nativeBuildInputs = [ cmake pkgconfig ];
+  buildInputs = [ boost nix ];
+}
```
`meta.license` should be added. In this case, because code from Nix is taken (LGPL 2.1).

Would it make sense to unpack the Nix source and copy it from there or does that make it too complex?

chkno replied:
License added.

Unpacking and copying from nix sources: That would probably be more more trouble than it's worth -- more code to do that than is copied over.  It's not critical that it be kept in perfect sync with nix -- these are convenience utility-style functions, not things that depend upon nix implementation details in a brittle way.  I plan to propose to the nix maintainers making these utility functions part of nix's exported API, but I wanted to move forward with this in parallel.  Ideally, this copied-from-nix code will go away soon.

Thanks!

chkno committed:
Add license

This is important because this contains some code copied from nix (as an
interim expediency until that functionality can be exported via nix's
API).  The license specified here must be compatible with this borrowing.
Select the same license that nix is released under: lgpl2Plus.

chkno committed:
Correct syntax for license specification

chkno committed:
Preserve type of rethrown exceptions

chkno committed:
Narrow the «not defined» check to just ThrownError

chkno committed:
Fix missing "using ThrownError"

chkno committed:
clang-format

chkno committed:
Support submodules (Fixes #13121)

Ma27 reviewed (commented):
First of all thanks a lot for working on this! The current implementation of `nixos-option` is indeed rather fragile and should be replaced entirely IMHO.

However I'd like to express a few concerns about the current approach:

* When we work on `nixos-option` we should also take a refactoring of [`nixops show-option`](https://nixos.org/nixops/manual/#command-nixops-show-option) into account: this command [uses strict evaluation](https://github.com/NixOS/nixops/blob/b5b805215b5071bfe7a1e256e660d977faf329be/nixops/deployment.py#L320-L334) to introspect attributes of the current deployment which happens to cause evaluation failures in several cases and doesn't provide a human-friendly output e.g. when querying an attribute set. So I guess that it would make sense to build a solution that can be easily used with `nixops show-option`.

* Not sure if evaluating the entire system config every time is worth the overhead since it needs to be re-implemented by several other tools (e.g. I built a wrapper for `evalModules` in [`nix-zsh-completions`](https://github.com/spwhitt/nix-zsh-completions/blob/master/_nixos-option)). Some months ago I discussed that with @fpletz and @elseym and we had the idea that a config activation should dump the current config (including details about where a value was declared) as JSON in `/run/current-system`, in that case tools like `nixops show-option`, `nixos-option`, their autocompletions and any other arbitrary tool would simply need to parse the JSON. Unfortunately I lacked time to build a proper implementation for that, but mostly out of interest, what would you think of that? :)

chkno commented:
Ma27 - Thanks for your comments!  I'm new to Nix.  I don't know what nixops is, so I will [read about nixops](https://nixos.org/nixops/manual/) and get back to you.

Mic92 commented:
@Ma27 Do you think this should be done as part of this pull request or can this be solved in future contributions?

lheckemann commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+            options_expr = nix::getArg(*arg, arg, end);
+        } else if (*arg == "--config_expr") {
+            config_expr = nix::getArg(*arg, arg, end);
+        } else if (!arg->empty() && arg->at(0) == '-') {
```
Maybe unrecognised options should be passed onto nix itself, e.g. to enable pure evaluation, set tarball ttl (when having URLs on the NIX_PATH for instance), or set `system` (to evaluate foreign-arch systems).

chkno replied:
I think this happens already by virtue of using `nix::LegacyArgs` and/or `nix::MixEvalArgs` above.  I observe that I can pass `-I nixpkgs=/some/other/path` to make it read the package definitions from elsewhere, even though there is no explicit handling of `-I` here.  (In fact, this is super-useful when used as `diff -u <(nixos-option --all) <(nixos-option --all -I nixpkgs=/some/other/path)`)

lheckemann commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+        auto const & attr = *i;
+        v = evaluateValue(ctx, &v);
+        if (attr.empty()) {
+            throw Error("empty attribute name in selection path '%s'", path);
```
nit: attribute names can be empty. Not sure we want to be encouraging that kind of behaviour though, so I'm fine with leaving this as is :)

chkno replied:
I was following `findAlongAttrPath` in `nix/src/libexpr/attr-path.cc`, which forbids empty attribute names.  Since these should probably agree, I'm leaving this check in.

lheckemann commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+        } else {
+            auto const & next = v.attrs->find(ctx->state->symbols.create(attr));
+            if (next == v.attrs->end()) {
+                throw Error("attribute '%s' in path '%s' not found", attr, path);
```
Not sure exactly why, but this error gets thrown with paths that contain `.`, e.g.:
```
$ nix eval '(with import <nixpkgs/nixos> {}; config.systemd.services."tinc.lugnet".serviceConfig.ExecStart)'
"/nix/store/v30a9csfl2w6wrj6lg24r09p946pvqy6-tinc-1.1pre17/bin/tincd -D -U tinc.lugnet -n lugnet -R --pidfile /run/tinc.lugnet.pid -d 0"
$ ./result/bin/nixos-option 'systemd.services."tinc.lugnet".serviceConfig.ExecStart'
error: attribute 'tinc.lugnet' in path 'systemd.services."tinc.lugnet".serviceConfig.ExecStart' not found
An error occurred while looking for attribute names. Are you sure that 'systemd.services."tinc.lugnet".serviceConfig.ExecStart' exists?
```

That said, I love that this _will_ presumably support attr names with dots in them!

chkno replied:
I don't have a `tinc.lugnet` service on my machine, but I do have a `environment.etc."host.conf".text`, which is a path with a dot in it.  Note that the interaction with shell quoting is somewhat confusing -- one must quote the paths to keep the shell from eating them:

```
$ ./result/bin/nixos-option environment.etc."host.conf".text
error: At 'conf' in path 'environment.etc.host.conf.text': Attribute not found

$ ./result/bin/nixos-option environment.etc.\"host.conf\".text
Value:
"multi on"

$ ./result/bin/nixos-option 'environment.etc."host.conf".text'
Value:
"multi on"
```

lheckemann replied:
Seems to be fixed in the current version, thanks!

lheckemann commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+    Value getSubOptions =
+        evaluateValue(ctx, findAlongAttrPath(*ctx->state, "type.getSubOptions", *ctx->autoArgs, option));
+    if (getSubOptions.type != tLambda) {
+        throw Error("Option's type.getSubOptions isn't a function");
```
I'm guessing this is a pretty rare/obscure error, but when it's encountered it would be helpful to know which option (attr path) this applies to.

chkno replied:
Done

lheckemann commented on nixos/modules/installer/tools/nixos-option/default.nix:
```diff
@@ -0,0 +1,10 @@
+  nativeBuildInputs = [ cmake pkgconfig ];
+  buildInputs = [ boost nix ];
+  meta = {
+    license = stdenv.lib.licenses.lgpl2Plus;
```
maintainer would probably be good here too

chkno replied:
Done

lheckemann reviewed (approved):
Wow, an impressive first contribution! I love it!

I've skimmed over the source, but won't claim to have reviewed it exhaustively. With:
- the adjustments I've commented
- a review from someone else (more familiar with nix's code, ideally!)
- and an entry in the release notes,

I'd be happy to merge this. A few more questions:

- Is there any particular reason not to have this as a standalone tool that can be built from nixpkgs without going through nixos, i.e. add it to `all-packages.nix`?

- re @Ma27's suggestion — I think this approach is nicer. Evaluating individual options isn't nearly as expensive as evaluating the whole system config, and I like the ability to look at system options before building the system. That said, this would also be an excellent base for creating that dump and having both ways available!

- @ma27 re `nixops show-option` — that should be fairly easy to implement on top of this with the help of the `--config_expr` and `--option_expr` options?

We've missed the branch-off for 19.09, but I'm already excited to have this as a feature in 20.03! :blush:

edolstra commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+    return buf.str();
+}
+
+std::string const appendPath(std::string const & prefix, std::string const & suffix)
```
For consistency with the Nix C++ convention, could you write `std::string const` as `const std::string`?

chkno replied:
Done

edolstra commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+        if (evaluated_type.type != tString) {
+            return false;
+        }
+        return evaluated_type.string.s == static_cast<std::string>("option");
```
Shouldn't this be `std::string(evaluated_type.string.s) == "option"`?

chkno replied:
I don't think it matters.  As long as one or the other of them is a std::string, I get an operator== that looks at content rather than pointer equality.  I picked casting the constant over casting the dynamic thing in hopes that the compiler would have a better chance at optimizing away any runtime cost.

But, I don't think this matters, so I switched it as you suggested.

edolstra commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+    std::string indentation;
+    std::string end;
+    LinePolicy policy;
+    bool write_since_sep;
```
`camelCase` for variable names please, e.g. `writeSinceSep`.

chkno replied:
Done

edolstra commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+// Stuff needed for evaluation
+struct Context
+{
+    Context(EvalState * state, Bindings * autoArgs, Value options_root, Value config_root)
```
To convey the fact that `state` and `autoArgs` cannot be null, I think references are better (i.e. `EvalState & state`).

chkno replied:
Done

edolstra commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+{
+    try {
+        if (std::holds_alternative<std::exception_ptr>(maybe_value)) {
+            std::rethrow_exception(std::get<std::exception_ptr>(maybe_value));
```
I think it's cleaner to write something like
```
if (ex = std::get_if<std::exception_ptr>(maybe_value))
    std::rethrow_exception(*ex);
```

chkno replied:
Done

edolstra commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+            out << v;
+        }
+    } catch (ThrownError & e) {
+        if (e.msg() == "The option `" + path + "' is used but not defined.") {
```
Can you add a comment in lib/modules.nix that if this message is changed, then `nixos-option` needs to be updated as well.

chkno replied:
Done

nbp commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+{
+    // Print this header on stderr rather than stdout because the old shell script
+    // implementation did.  I don't know why.
+    std::cerr << "This attribute set contains:\n";
```
To be used by other tools without the need of other tools to filter it out.
I guess this can be remove.

chkno replied:
By "I guess this can be removed." - do you mean removing this header entirely, or removing the special stderr-nature of it?

I think you meant the special stderr-treatment can be removed, so I did that.

lheckemann replied:
I thought it was just the comment. I don't think it's a big disaster either way though :)

nbp reviewed (approved):
Thank you!

While I do not like the verbosity of C++ compared to Nix, this code has better handling of corner cases and error messages than the previous script. However, I would suggest to implement as much as possible in Nix, instead of C++, when possible.

chkno committed:
Switch from east const to west const

For consistency with the Nix C++ convention.

:~(

chkno committed:
Add note: Keep error message in sync with nixos-option

chkno committed:
Always say which path component had trouble

chkno committed:
Per reviewer request, cast the other side.

I don't think this matters.  As long as one or the other of these is
a std::string, I get an operator== that looks at content rather than
pointer equality.  I picked casting the constant over casting the dynamic
thing in hopes that the compiler would have a better chance at optimizing
away any runtime cost.

Deferring to reviewer.

chkno committed:
Use std::get_if

chkno committed:
Hold state and autoArgs by reference

Switch from convention "appease clang-tidy --checks='*'" to
"References are like non-nullptr pointers".  The clang-tidy check
"google-runtime-references" complains about non-const reference
arguments, but this is not a convention used in Nix.

chkno committed:
Pass Context by reference

Switch from convention "appease clang-tidy --checks='*'" to
"References are like non-nullptr pointers".  The clang-tidy check
"google-runtime-references" complains about non-const reference
arguments, but this is not a convention used in Nix.

chkno committed:
Pass values by reference

Switch from convention "appease clang-tidy --checks='*'" to
"References are like non-nullptr pointers".  The clang-tidy check
"google-runtime-references" complains about non-const reference
arguments, but this is not a convention used in Nix.

chkno committed:
Explain why header goes on stderr

chkno committed:
Don't print header on stderr

Automated consumers can use 'sed 1d' or similar to remove this header.

This probably makes this output *easier* to consume correctly.  Having
this header show up in consumers' terminal or log output is probably not
useful, but hiding it without hiding all error messages would have been
more troublesome that just stripping it from stdout.

I.e., previously, unsophisticated use would show undesired output:
  $ some-other-tool
  This attribute set contains:
  This attribute set contains:
  This attribute set contains:
  This attribute set contains:
  <Actual some-other-tool output>

The simplest way to hide this undesired output would have been
nixos-option ... 2>/dev/null, which would hide all error messages.
We do not wish to encourage that.

Correct use would have been something like:
  nixos-option ... 2> >( grep --line-buffered -v 'This attribute set contains:')

After this change, correct use is simpler:
  nixos-option ... | sed 1d
or
  nixos-option ... | sed '1/This attribute set contains:/d'
if the caller don't know if this invocation of nixos-option will yield
an attribute listing or an option description.

chkno committed:
snake_case -> camelCase

chkno committed:
Only print example when there is one

chkno committed:
Add maintainer

chkno commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,599 @@
+    out << "\n\nType:\n";
+    printAttr(ctx, out, "type.description", option);
+
+    out << "\n\nExample:\n";
```
Done

chkno commented:
> While I do not like the verbosity of C++ compared to Nix, this code has better handling of corner cases and error messages than the previous script. However, I would suggest to implement as much as possible in Nix, instead of C++, when possible.

I didn't do this because I was afraid of the performance cost.  I want the `diff <(nixos-option --all) <(nixos-option --all ...)` use case to be fast enough that it doesn't cause humans to context-switch while waiting for it.  It's already 10-14 seconds depending on the speed of the machine, which is already quite long for holding a human's undivided attention.

chkno committed:
(clang-format for has_example)

chkno committed:
Support aggregate types attrsOf and listOf

Mic92 commented on nixos/modules/installer/tools/nixos-option/nixos-option.cc:
```diff
@@ -0,0 +1,618 @@
+    }
+}
+
+bool has_example(Context & ctx, Value & option)
```
nitpick:

```suggestion
bool hasExample(Context & ctx, Value & option)
```

chkno replied:
Fixed.  Thanks!

chkno committed:
snake_case -> camelCase

lheckemann commented:
I think that with an entry in the release notes (`nixos/doc/manual/release-notes/rl-2003.xml`), this should be good to merge. I can do that if you want?

globin committed:
release-notes 20.03: added notice for rewritten nixos-option

deliciouslytyped commented:
I think it would be nice to add in the release note that this also improves correctness over the bash script, which I had problems with. :)

aviallon commented:
Does this new version support flakes?

vcunat commented:
Merged over five years ago and you call it new?

phanirithvij commented:
I assume aviallon is talking about the new new version, https://github.com/NixOS/nixpkgs/pull/369151
